Features/update date input #157785070

chrisdolendo opened the pull request:
This branch contains:
Features/update date input #157785070

Caritas Feedback Mini-Epic stories are currently branched off from `lease-up-preprod` because they will be released shortly after Lease Up Release 1 (found in lease-up-preprod) is brought to production (and should not contain post-lease-up release 1 stories). 

Every time a caritas story is merged into the `caritas-feedback-release` branch we will update `master` with `caritas-feedback-release` so master has caritas code. Once this mini-epic is complete, we will bring it to production.

chrisdolendo committed:
[#157785070] - Remove datepicker usage sitewide

akegan commented:
@chrisdolendo FYI i just updated the title of your PR to be your feature again

akegan commented on app/javascript/components/applications/application_form/PaperApplicationForm.js:
```diff
@@ -79,7 +80,8 @@ class PaperApplicationForm extends React.Component {
 
-    this.setState({ loading: false })
+    this.setState({ submittedValues, loading: true, failed: false })
+    await onSubmit(submitType, submittedValues, application, listing)
```
isn't onSubmit calling this submitShortForm? why do we need to call it again?

akegan commented on app/javascript/components/applications/application_form/PaperApplicationForm.js:
```diff
@@ -79,7 +80,8 @@ class PaperApplicationForm extends React.Component {
     }
 
-    this.setState({ loading: false })
+    this.setState({ submittedValues, loading: true, failed: false })
```
looks like loading is set to true at the top of this function, do we need to set it again? are we setting it to be false anywhere?

akegan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -15,15 +16,46 @@ let mailingAddressFieldMap = {
+    return { error: !!formApi.errors[field] }
+  else
+    return null
+}
```
are these functions general enough that they could be pulled into a util file or something?

akegan replied:
I still think that some of these should be pulled out into a util function, but that can probably wait until the next PR

fddayan replied:
yes, actually I'm working on a different story with validation and pulling this out

akegan replied:
on master or on caritas?

fddayan replied:
caritas

akegan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -33,15 +65,15 @@ const PrimaryApplicantSection = ({formApi, editValues }) => {
                 <div className="small-4 columns">
                   <label>First Name <span className="checkbox-block_note no-margin">(required)</span></label>
-                  <Text required="true" field="firstName" />
+                  <Text field="first_name" />
```
is field tied to the variable stored in formAPI? do we want to change all of them to snake case?

chrisdolendo replied:
@akegan This is and all the fields are actually in snake case, it's because we do a field mapping from SOQL to ShortForm Custom API and back to stay consistent. So instead of firstName it's first_name, dateOfBirth is date_of_birth, lastName is last_name in the React world. And we have a service that does the back and forth mapping form SOQL<-> ShortForm Custom API

akegan replied:
other forms that I was looking at have field='firstName' though. Should we update all of them?

chrisdolendo replied:
Oh wait I was totally looking at the wrong thing. I'm looking at the code now and I think you're right and actually caught a bug when you actually submit 👍 Thank for flagging, debugging now 😸

akegan reviewed (commented):
Just some questions!

chrisdolendo committed:
[#157785070] - Updated date input for primary applicant and household member

- Removed use of datepicker
- Created error messages/validation on field touch and submit
- Scrolls user up to the stop when form is invalid on submit

chrisdolendo committed:
[#157785070] - Updated tests

akegan commented:
merge conflicts 😢

chrisdolendo committed:
[#157785070] - Merged with caritas-feedback-release branch, fixed conflicts

callaghanc committed:
[#157785070] Adds fallback for DOB field for browsers that don't support date input type

We are using the HTML5 "date" input field type on the DOB fields. However, not all browsers support this input type. Notably, Safari and IE do not support it. DAHLIA supports IE10+ and Safari. So this commit adds attributes to the DOB fields to ensure that the fields produce the same values in browsers that do not support the "date" input type. Please see this MDN doc for more info: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date.

callaghanc commented:
@akegan i added some updates here, so let's make sure that you also add your approving review before we call this delivered

akegan commented:
validation is kinda weird/not great on safari. This is what pops up on submit
 
it also randomly changes between saying applicant has to be >18, and not saying anything til you submit
![image](https://user-images.githubusercontent.com/11825994/43872432-84edbf9e-9b36-11e8-8b25-934410195e0e.png)

akegan commented:
![image](https://user-images.githubusercontent.com/11825994/43872478-b4f24336-9b36-11e8-9c4e-6e40ee4ac4de.png)

![image](https://user-images.githubusercontent.com/11825994/43872493-c93a5f54-9b36-11e8-90ff-8b4d1b75b706.png)

akegan commented:
it also looks like it should be MM-DD-YYYY from the acceptance criteria, but it's reversed on safari

akegan committed:
[#157785070] Add inline validation for valid date

akegan commented:
Just pushed a commit that provides inline validation for valid date format and dob that works on chrome and safari. 

It looks like the order of year,month,day on safari is locked down based on the way the date is stored as a value with the html5 field

fddayan commented on app/javascript/components/applications/application_form/HouseholdMemberForm.js:
```diff
@@ -22,13 +20,12 @@ const HouseholdMemberForm = ({ i }) => {
-              showYearDropdown
-              dropdownMode="select"
-              field={`householdMembers.${i}.DOB`} />
+            <label>Date of Birth</label>
```
should we include `(required)` in the label as it was previously?

fddayan commented on app/javascript/components/applications/application_form/HouseholdMemberForm.js:
```diff
@@ -22,13 +20,12 @@ const HouseholdMemberForm = ({ i }) => {
+            <label>Date of Birth</label>
+            <Text
+              field={`householdMembers.${i}.DOB`}
+              type="date"
```
here to should we include `required={true}` ? or it is not required any more?

chrisdolendo replied:
`required` uses the browser validation which messes with ours so we are going to reimplement it our way in https://www.pivotaltracker.com/story/show/159244905

chrisdolendo committed:
[#157785070] - Remove console.log

chrisdolendo commented:
paired with @akegan fixes look good! thanks! Delivering the story

fddayan commented on app/javascript/components/applications/application_form/PrimaryApplicantSection.js:
```diff
@@ -61,14 +101,16 @@ const PrimaryApplicantSection = ({formApi, editValues }) => {
-                  dropdownMode="select"
-                  field="DOB" />
+                <div className={classNames('form-group', errorClassName(formApi,'DOB'))}>
+                  <label className='form-label'>Date of Birth</label>
```
is `required` missing here too?

fddayan reviewed (dismissed):
A few questions about the field being `required`

chrisdolendo commented:
@fddayan Re required Q's:`required` uses the browser validation which messes with our React Form validation because it triggers the browser's html5 validation-- so we are going to reimplement it our way using `validateErrors` in https://www.pivotaltracker.com/story/show/159244905.

akegan committed:
[#157785070] Add better date input validation

akegan committed:
Merge branch 'features/update-date-input-#157785070' of https://github.com/Exygy/sf-dahlia-lap into features/update-date-input-#157785070

fddayan reviewed (approved):
👍

chrisdolendo commented:
@akegan :+1: your addition looks good to me
